workflow-core/src/commonMain/kotlin/com/squareup/workflow1/WorkflowTracer.kt

```diff
@@ -0,0 +1,36 @@
+package com.squareup.workflow1
+
+/**
+ * This is a very simple tracing interface that can be passed into a workflow runtime in order
+ * to inject span tracing throughout the workflow core and runtime internals.
+ */
+public interface WorkflowTracer {
+  public fun beginSection(label: String): Unit
+  public fun endSection(): Unit
+}
+
+/**
+ * Convenience function to wrap [block] with a trace span as defined by [WorkflowTracer].
+ * Only calls [label] if there is an active [WorkflowTracer] use this for any label other than
+ * a constant.
+ */
+public inline fun <T> WorkflowTracer?.trace(label: () -> String, block: () -> T): T {
+  val optimizedLabel = if (this !== null) {
+    label()
+  } else {
+    ""
+  }
+  return trace(optimizedLabel, block)
+}
+
+/**
+ * Convenience function to wrap [block] with a trace span as defined by [WorkflowTracer].
+ */
+public inline fun <T> WorkflowTracer?.trace(label: String, block: () -> T): T {
+  this?.beginSection(label)
+  try {
```

Comment:
hmm good question. I'd have to look at the decompiled byte code I guess

Reply:
given that its `inline` if I branch it to only use the `try / finally` when the tracer is non-null, what is the other concern you referred to? I think this would be fine as it would be inlined and then the unused branch shaken out of the code?

Reply:
> then the unused branch shaken out of the code?

There's no shaking out at compile time as whether WorkflowTracer is null or not is determined at runtime.

I'm not aware of any try / finally cost, however the resulting bytecode here shows we're checking whether WorkflowTracer is null twice, which also means we're reaching into the workflow tracer field twice.

This avoids the double null check:

```
public inline fun <T> WorkflowTracer?.trace(label: String, block: () -> T): T {
  return if (this == null) {
    block()
  } else {
    beginSection(label)
    try {
      block()
    } finally {
      endSection()
    }
  }
}
```

Reply:
Done.